### PR TITLE
Improve errors for non-existing variables

### PR DESCRIPTION
**Pull request by fpagnoux**
Connected to #398

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## 8.0.0
+
+* Raise more explicit error when an invalid test case is given
+* Raise more friendly error when trying to calculate a variable which doesn't exist
+* _Breaking change_ : Python exceptions will now be raised when the test case in invalid. Before, only a byriani error was returned. Reusers must therefore adapt their exception handling.
```
Comment (cbenz) [inline CHANGELOG.md]: No space before ':' in English :)

File: openfisca_core/json_to_test_case.py [hunk 2]
```diff
@@ -1,85 +1,100 @@
 # -*- coding: utf-8 -*-
 
-import itertools
-
 from . import conv
+from copy import deepcopy
```
Comment (cbenz) [inline openfisca_core/json_to_test_case.py]: Separate standard library imports from local imports (like it was before your commit with `itertools`)

File: openfisca_core/json_to_test_case.py [hunk 3]
```diff
@@ -1,85 +1,100 @@
...
-                )
-
-    def get_role_parsing_dict(entity):
-        if entity.is_person:
-            return {}
+                    )
+                )(value)
+
+        if error is not None:
+            raise ValueError(u"Invalid description of {}: {}. Error: {}".format(entity_class.key, entity_json, error))
```
Comment (cbenz) [inline openfisca_core/json_to_test_case.py]: Exception message must be encoded in UTF-8

```python
raise ValueError(u'...{}...'.format(foo).encode('utf-8'))
```

File: openfisca_core/json_to_test_case.py [hunk 4]
```diff
@@ -1,85 +1,100 @@
...
+        entity_json[key] = value
+
+    def check_variable(value, key):
+        column = tax_benefit_system.column_by_name[key]
+        if column.entity != entity_class:
+            raise ValueError(u"Variable {} is defined for entity {}. It cannot be set for entity {}.".format(key, column.entity.key, entity_class.key))
```
Comment (cbenz) [inline openfisca_core/json_to_test_case.py]: Exception message must be encoded in UTF-8

File: openfisca_core/json_to_test_case.py [hunk 5]
```diff
@@ -1,85 +1,100 @@
...
+        value, error = column.json_to_python(value)
+        if error is not None:
+            raise ValueError(u"Invalid value {} for variable {}. Error: {}".format(value, key, error))
```
Comment (cbenz) [inline openfisca_core/json_to_test_case.py]: Exception message must be encoded in UTF-8

File: openfisca_core/json_to_test_case.py [hunk 6]
```diff
@@ -1,85 +1,100 @@
...
-                            )),
-                        drop_none_values = True,
-                        ),
-                    drop_none_items = True,
-                    ),
-                conv.default([]),
-                )
-            for entity in tax_benefit_system.entities
-            }
+            from .taxbenefitsystems import VariableNotFound
```
Comment (cbenz) [inline openfisca_core/json_to_test_case.py]: If it's really necessary to have this import here, please add a comment to explain why

File: openfisca_core/json_to_test_case.py [hunk 7]
```diff
@@ -1,85 +1,100 @@
...
+            - An entity is not recognized
+            - An entity role is not recognized
+            - A variable is declared for an entity it is not defined for (e.g. salary for a family)
+    """
+
+    test_case = deepcopy(test_case)  # Avoid side-effects on other references to test_case
+    entity_classes = {entity_class.plural: entity_class for entity_class in tax_benefit_system.entities}
+    for entity_type_name, entities in test_case.iteritems():
+        if entity_classes.get(entity_type_name) is None:
+            raise ValueError(u"Invalid entity name: {}".format(entity_type_name))
```
Comment (cbenz) [inline openfisca_core/json_to_test_case.py]: Exception message must be encoded in UTF-8

File: openfisca_core/json_to_test_case.py [hunk 8]
```diff
@@ -1,85 +1,100 @@
...
+            conv.make_item_to_singleton(),
+            conv.test_isinstance(list),
+            conv.uniform_sequence(
+                conv.test_isinstance(dict),
+                drop_none_items = True,
+                ),
+            conv.function(set_entities_json_id),
+            )(entities)
+        if error is not None:
+            raise ValueError(u"Invalid list of {}: {}. Error: {}".format(entity_type_name, entities, error))
```
Comment (cbenz) [inline openfisca_core/json_to_test_case.py]: Exception message must be encoded in UTF-8

File: openfisca_core/taxbenefitsystems.py [hunk 9]
```diff
@@ -20,7 +22,14 @@
...
 class VariableNotFound(Exception):
-    pass
+    def __init__(self, variable_name, tax_benefit_system):
+        country_package_name, country_package_version = tax_benefit_system.get_package_metadata()
+        message = (
+            u"You tried to calculate or to set a value for variable '{0}', but it was not found in the loaded tax benefit system ({1}@{2}). "
+            u"Are you sure you spelled '{0}' correctly? "
+            u"If this code used to work and suddenly does not, this is most probably linked to an update of the tax benefit system. Look at its changelog to learn about renames and removals and update your code. If it is an official package, it is probably available on <https://github.com/openfisca/{1}/blob/master/CHANGELOG.md>."
+            ).format(variable_name, country_package_name, country_package_version)
+        Exception.__init__(self, message)
```
Comment (cbenz) [inline openfisca_core/taxbenefitsystems.py]: Exception message must be encoded in UTF-8

```python
Exception.__init__(self, message.encode('utf-8'))
```

File: openfisca_core/json_to_test_case.py (hunk 2 above)
Comment (fpagnoux) [inline openfisca_core/json_to_test_case.py]: 👍

**Pull request merged**